packages/testing/src/chai-retry-plugin/chai-retry-plugin.ts

```diff
@@ -0,0 +1,187 @@
+import Chai, { AssertionError } from 'chai';
+import { timeout as addTimeoutToPromise, sleep } from 'promise-assist';
+
+import type {
+    AssertionMethod,
+    FunctionToRetry,
+    AssertionStackItem,
+    RetryAndAssertProps,
+    RetryOptions,
+    PromiseLikeAssertion,
+    AssertionPropertyKeys,
+} from './types';
+
+const { Assertion, expect } = Chai;
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Chai {
+        interface Assertion {
+            /**
+             * Allows to retry the function passed to `expect` and assert the result until retries ended or timeout exceeded
+             * @param options Settings for retry logic:
+             * - `timeout`: The maximum duration in milliseconds to wait before failing the retry operation.
+             * - `retries`: The number of times to retry the function before failing.
+             * - `delay`: The delay in milliseconds between retries.
+             */
+            retry(options?: RetryOptions): PromiseLikeAssertion;
+        }
+    }
+}
+
+const retryFunctionAndAssertResult = async ({ functionToRetry, options, assertionStack }: RetryAndAssertProps) => {
+    const { retries, delay } = options;
+    let retriesCount = 0;
+
+    while (retriesCount < retries) {
+        try {
+            retriesCount++;
+            const result = await functionToRetry();
+            const assertion = expect(result);
+            let negationIsApplied = false;
+
+            for (const { isNegate, method, args = [], key } of assertionStack) {
+                if (isNegate) {
+                    negationIsApplied = true;
```

Comment:
Should this change `negationIsApplied` to the opposite value instead of just set true?
what about `not.not.not`?

Reply:
should `and` reset negation?
e.g.
```
be.not.greater(4).and.be.above(10);
```

Reply:
@olehrakwix no, in native Chai's logic if you set `not` once it is propagated to the rest of assertion, another `not` in chain will change nothing

Attaching screenshot with examples:

<img width="654" alt="image" src="https://user-images.githubusercontent.com/124504800/229993022-deb05b7c-91e3-4ddb-b319-08d6607725b3.png">

Reply:
@olehrakwix removed custom handling of negated assertions